pkg/apis/resource/validation/validation.go

```diff
@@ -17,172 +17,260 @@ limitations under the License.
+func validateDeviceClaim(deviceClaim *resource.DeviceClaim, fldPath *field.Path, stored bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	requestNames := gatherRequestNames(deviceClaim)
+	allErrs = append(allErrs, validateSet(deviceClaim.Requests, resource.DeviceRequestsMaxSize,
+		func(request resource.DeviceRequest, fldPath *field.Path) field.ErrorList {
+			return validateDeviceRequest(request, fldPath, stored)
+		},
+		func(request resource.DeviceRequest) (string, string) {
+			return request.Name, "name"
+		},
+		fldPath.Child("requests"))...)
+	allErrs = append(allErrs, validateSlice(deviceClaim.Constraints, resource.DeviceConstraintsMaxSize,
+		func(constraint resource.DeviceConstraint, fldPath *field.Path) field.ErrorList {
+			return validateDeviceConstraint(constraint, fldPath, requestNames)
+		}, fldPath.Child("constraints"))...)
+	allErrs = append(allErrs, validateSlice(deviceClaim.Config, resource.DeviceConfigMaxSize,
+		func(config resource.DeviceClaimConfiguration, fldPath *field.Path) field.ErrorList {
+			return validateDeviceClaimConfiguration(config, fldPath, requestNames)
+		}, fldPath.Child("config"))...)
+	return allErrs
+}
+
+func gatherRequestNames(deviceClaim *resource.DeviceClaim) sets.Set[string] {
+	requestNames := sets.New[string]()
+	for _, request := range deviceClaim.Requests {
+		requestNames.Insert(request.Name)
+	}
+	return requestNames
+}
+
+func validateDeviceRequest(request resource.DeviceRequest, fldPath *field.Path, stored bool) field.ErrorList {
+	allErrs := validateRequestName(request.Name, fldPath.Child("name"))
+	if request.DeviceClassName == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("deviceClassName"), ""))
 	} else {
-		for _, msg := range pathvalidation.IsValidPathSegmentName(ref.Kind) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("kind"), ref.Kind, msg))
+		allErrs = append(allErrs, validateDeviceClassName(request.DeviceClassName, fldPath.Child("deviceClassName"))...)
+	}
+	allErrs = append(allErrs, validateSlice(request.Selectors, resource.DeviceSelectorsMaxSize,
+		func(selector resource.DeviceSelector, fldPath *field.Path) field.ErrorList {
+			return validateSelector(selector, fldPath, stored)
+		},
+		fldPath.Child("selectors"))...)
+	switch request.AllocationMode {
+	case resource.DeviceAllocationModeAll:
+		if request.Count != 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("count"), request.Count, "a non-zero count is not valid for allocationMode="+string(request.AllocationMode)))
+		}
+	case resource.DeviceAllocationModeExactCount:
+		if request.Count <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("count"), request.Count, "must be greater than zero"))
 		}
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("allocationMode"), request.Count, fmt.Sprintf("must be one of: %s", strings.Join([]string{string(resource.DeviceAllocationModeAll), string(resource.DeviceAllocationModeExactCount)}, ", "))))
```

Comment:
fireld.NotSupported() handles this for you

Reply:
I swear I made that change in a branch, but as part of a larger effort, and now I cannot find it.  I will try to revive it, but aync to this PR